crates/ethereum/src/lib.rs

```diff
@@ -1,111 +1,354 @@
+            tracing::error!(code, message, "Ethereum call failed");
+            anyhow::bail!(code);
+        }
+
+        let response: serde_json::Value = res.json().await?;
+        Ok(response["result"].clone())
     }
 }
 
-impl TryFrom<&ethers::types::Log> for TransactionOrigin {
-    type Error = anyhow::Error;
-
-    fn try_from(log: &ethers::types::Log) -> Result<Self, Self::Error> {
-        let hash = log.transaction_hash.context("missing transaction hash")?;
-        let hash = EthereumTransactionHash(hash);
-        let index = log
-            .transaction_index
-            .context("missing transaction index")?
-            .as_u64();
-        let index = EthereumTransactionIndex(index);
-        Ok(Self { hash, index })
+#[async_trait::async_trait]
+impl EthereumApi for EthereumClient {
+    async fn get_starknet_state(&self, address: &H160) -> anyhow::Result<EthereumStateUpdate> {
+        let hash = self.get_latest_block_hash().await?;
+        let hash = format!("0x{}", hex::encode(hash.as_bytes()));
+        let addr = format!("0x{}", hex::encode(address.as_bytes()));
+        Ok(EthereumStateUpdate {
+            global_root: self
+                .call_starknet_contract(&hash, &addr, "stateRoot()")
+                .await
+                .and_then(|value| get_h256(&value))
+                .and_then(get_felt)
+                .map(StateCommitment)?,
+            block_hash: self
+                .call_starknet_contract(&hash, &addr, "stateBlockHash()")
+                .await
+                .and_then(|value| get_h256(&value))
+                .and_then(get_felt)
+                .map(BlockHash)?,
+            block_number: self
+                .call_starknet_contract(&hash, &addr, "stateBlockNumber()")
+                .await
+                .and_then(|value| get_u256(&value))
+                .and_then(get_number)?,
+        })
+    }
+
+    async fn get_chain(&self) -> anyhow::Result<EthereumChain> {
+        let id = self
+            .call_ethereum(serde_json::json!({
+                "jsonrpc": "2.0",
+                "method": "eth_chainId",
+                "params": [],
+                "id": 0
+            }))
+            .await
+            .and_then(|value| get_u256(&value))?;
+        Ok(match id {
+            x if x == U256::from(1u32) => EthereumChain::Mainnet,
+            x if x == U256::from(5u32) => EthereumChain::Goerli,
+            x => EthereumChain::Other(x),
+        })
     }
 }
 
-impl TryFrom<&ethers::types::Log> for EthOrigin {
-    type Error = anyhow::Error;
+fn encode_ethereum_call_data(signature: &[u8]) -> String {
+    let mut output: [u8; 32] = Default::default();
+    keccak_hash::keccak_256(signature, &mut output[..]);
+    format!("0x{}", hex::encode(&output[0..4]))
+}
+
+fn get_h256(value: &serde_json::Value) -> anyhow::Result<H256> {
+    use std::str::FromStr;
+    value
+        .as_str()
+        .map(lpad64)
+        .and_then(|val| H256::from_str(&val).ok())
```

Comment:
```suggestion
        .ok_or(anyhow::anyhow!("Failed to parse H256"))
```

Reply:
already there, please check next commits

Reply:
Ah sorry, think I grabbed the wrong line. I just meant the phrasing is slightly wrong for this function since its parsing and not fetching. Same for the other.